docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,698 @@
+   // roots of each row and col of the erasure data).
+   dah := da.NewDataAvailabilityHeader(eds)
+
+   // We use the block data struct to pass the square size and calculated data
+   // root to tendermint.
+   blockData.Hash = dah.Hash()
+   blockData.OriginalSquareSize = squareSize
+
+   // tendermint doesn't need to use any of the erasure data, as only the
+   // protobuf encoded version of the block data is gossiped.
+   return abci.ResponsePrepareProposal{
+       BlockData: &blockData,
+   }
+}
+```
+
+#### ParsedTxs
+
+The first major change is that we are making use of an intermediate data structure. It contains fields that are progressively and optionally used during the malleation process. This makes it easier to keep track of malleated transactions and their messages, prune transactions in the case that we go over the max square size, cache the decoded transactions avoiding excessive deserialization, and add metadata to malleated transactions after we malleate them. All while preserving the original ordering (from the prioritized mempool) of the transactions.
+
+```go
+// parsedTx is an interanl struct that keeps track of potentially valid txs and
+// their wire messages if they have any.
+type parsedTx struct {
+   // the original raw bytes of the tx
+   rawTx []byte
+   // tx is the parsed sdk tx. this is nil for all txs that do not contain a
+   // MsgWirePayForData, as we do not need to parse other types of of transactions
+   tx signing.Tx
+   // msg is the wire msg if it exists in the tx. This field is nil for all txs
+   // that do not contain one.
+   msg *types.MsgWirePayForData
+   // malleatedTx is the transaction after the malleation process is performed. This is nil until that process has been completed for viable transactions.
+   malleatedTx coretypes.Tx
+}
+```
+
+```go
+func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+   // parse the txs, extracting any MsgWirePayForData and performing basic
+   // validation for each transaction. Invalid txs are ignored. Original order
+   // of the txs is maintained.
+   parsedTxs := parseTxs(app.txConfig, req.BlockData.Txs)
+   ...
+}
+```
+
+#### Square Estimation
+
+Using some of the non-interactive defaults code above, we can quickly calculate the number of shares required by each type of block data. As discussed in the docs of this function, the goal here is not necessarily to get a perfect count of the shares that are being used, but we do need to know roughly what square size is needed. When estimating, we should round up in square size.
+
+```go
+// FitsInSquare uses the non interactive default rules to see if messages of
+// some lengths will fit in a square of size origSquareSize starting at share
+// index cursor. See non-interactive default rules
+// https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#non-interactive-default-rules
```

Comment:
Do we really want to link to spec here and in the code? If yes, then please use permalinks.

Reply:
:+1: 

removed this specific reference, but used permalinks to the others 185e1574c1f746d8bd2067f165170e1275f0d364